File: src/main/java/com/pygostylia/osprey/Velocity.java
package com.pygostylia.osprey;

import java.io.IOException;

/*
The protocol representation of velocity is as a triple of Shorts,
representing velocity along the three axes in units of
1/8000 block / 50ms

Internally, I represent velocity as a float blocks / second, and
convert on-demand when encoding to the protocol layer.
 */
public record Velocity(float x, float y, float z) {
    public static Velocity zero() {
        return new Velocity(0, 0, 0);
    }

    public void write(MinecraftOutputStream os) throws IOException {
        os.writeShort(blockPerSecondToProtocol(x));
        os.writeShort(blockPerSecondToProtocol(y));
        os.writeShort(blockPerSecondToProtocol(z));
    }

    public static Velocity directionMagnitude(EntityPosition entityPosition, float speed) {
        var x = -Math.sin(entityPosition.yawRadians());
        var z = Math.cos(entityPosition.yawRadians());
        var yH = -Math.sin(entityPosition.pitchRadians());
        var yXZ = Math.cos(entityPosition.pitchRadians());
        x *= yXZ;
        z *= yXZ;
        return new Velocity(
                (float) x * speed,
                (float) yH * speed,
                (float) z * speed
        );
    }

    private static float protocolToBlockPerSecond(short protocol) {
        return protocol / 400f;
    }

    private static short blockPerSecondToProtocol(float blockPerSecond) {
        return (short) (blockPerSecond * 400f);
    }

    public Velocity divide(float dividend) {
        return new Velocity(x / dividend, y / dividend, z / dividend);
    }

    public Velocity add(float x, float y, float z) {
        return new Velocity(this.x + x, this.y + y, this.z + z);
    }

    public Velocity offsetGravity(float y) {
        return this.add(0, y, 0);
    }
}

File: src/main/java/com/pygostylia/osprey/Velocity.kt
package com.pygostylia.osprey

import com.pygostylia.osprey.streams.MinecraftOutputStream
import java.io.IOException
import kotlin.math.*

/*
The protocol representation of velocity is as a triple of Shorts,
representing velocity along the three axes in units of
1/8000 block / 50ms

Internally, I represent velocity as a float blocks / second, and
convert on-demand when encoding to the protocol layer.
 */
data class Velocity(val x: Float, val y: Float, val z: Float) {
    @Throws(IOException::class)
    fun write(os: MinecraftOutputStream) {
        os.writeShort(blockPerSecondToProtocol(x))
        os.writeShort(blockPerSecondToProtocol(y))
        os.writeShort(blockPerSecondToProtocol(z))
    }

    fun divide(dividend: Float): Velocity {
        return Velocity(x / dividend, y / dividend, z / dividend)
    }

    fun add(x: Float, y: Float, z: Float): Velocity {
        return Velocity(x + x, y + y, z + z)
    }

    fun offsetGravity(y: Float): Velocity {
        return add(0f, y, 0f)
    }

    companion object {
        @JvmStatic
        fun zero(): Velocity {
            return Velocity(0f, 0f, 0f)
        }

        @JvmStatic
        fun directionMagnitude(entityPosition: EntityPosition, speed: Float): Velocity {
            var x = -sin(entityPosition.yaw)
            var z = cos(entityPosition.yaw)
            val yH = -sin(entityPosition.pitch)
            val yXZ = cos(entityPosition.pitch)
            x *= yXZ
            z *= yXZ
            return Velocity(
                x.toFloat() * speed,
                yH.toFloat() * speed,
                z.toFloat() * speed
            )
        }

        private fun protocolToBlockPerSecond(protocol: Short): Float {
            return protocol / 400f
        }

        private fun blockPerSecondToProtocol(blockPerSecond: Float): Int {
            return (blockPerSecond * 400f).toInt()
        }
    }
}